tls/s2n_handshake_type.c

```diff
@@ -57,6 +57,7 @@ S2N_RESULT s2n_handshake_type_set_tls13_flag(struct s2n_connection *conn, s2n_tl
     RESULT_ENSURE_REF(conn);
     RESULT_ENSURE(s2n_connection_get_protocol_version(conn) >= S2N_TLS13, S2N_ERR_HANDSHAKE_STATE);
     conn->handshake.handshake_type |= flag;
+    RESULT_GUARD(s2n_conn_choose_state_machine(conn, S2N_TLS13));
```

Comment:
Would there be a problem if a user called `s2n_handshake_type_set_tls13_flag`, setting the state machine to TLS 1.3, and then called `s2n_handshake_type_reset`? Should the state machine be reset to S2N_STATE_MACHINE_INITIAL in `s2n_handshake_type_reset`?

Reply:
Sorry, these aren't public APIs...

Reply:
Yeah I actually considered putting this in a couple of our public APIs, like s2n_append_psk() and s2n_recv_early_data(). We technically want to set which state machine we're using as early as possible. But I'm worried about putting this under the control of the user... I dunno, I'm still thinking about this.

Reply:
Can't you still negotiate TLS1.2 even if you call s2n_append_psk or s2n_recv_early_data? We should only be setting it if we're absolutely sure which state machine-- so we've negotiated a version or we've set a tls1.3-only flag.

Reply:
I don't know if you can still negotiate TLS1.2 after calling those APIs... 🤔 I don't technically know what would prevent you from doing so...